tlatools/org.lamport.tlatools/src/tlc2/tool/impl/DebugTool.java

```diff
@@ -123,20 +124,28 @@ public boolean isValidAssumption(final ExprNode assumption) {
 	}
 
 	@Override
-	public boolean isValid(Action act, TLCState state) {
+	public InvariantViolation findFalsifyingContext(Action invariant, TLCState s0, TLCState s1) {
+		// Need this override for breakpoints to work
+		return isValid(invariant, s0, s1)
+			? null
+			: new InvariantViolation(invariant.pred, Context.Empty);
+	}
+
+	@Override
+	public boolean isValid(Action act, TLCState s0, TLCState s1) {
```

Comment:
I had to rework the way `DebugTool` overrides `isValid` or else there are hangs and failures in the debugger tests. I _think_ this new approach is better because
 - `isValid(Action, TLCState)` (with 1 state) is implemented in terms of this method
 - so overriding `isValid(Action, TLCState, TLCState)` suffices to handle both overloads

But there could be something I'm missing.

Reply:
I don't see any issues with the proposed change.

Note: You can use `DebugTool#noDebug` to obtain an instance of `ITool`, which evaluates expressions outside of the debugger's control.